language/stdlib/modules/VASP.move

```diff
@@ -58,73 +59,97 @@ module VASP {
     // To-be parent-vasp called functions
     ///////////////////////////////////////////////////////////////////////////
 
-    public fun create_parent_vasp_credential(
+    public fun publish_parent_vasp_credential(
+        association: &signer,
+        vasp: &signer,
         human_name: vector<u8>,
         base_url: vector<u8>,
         compliance_public_key: vector<u8>
-    ): ParentVASP {
+    ) {
+        Association::assert_is_root(association);
         Transaction::assert(Signature::ed25519_validate_pubkey(copy compliance_public_key), 7004);
-        ParentVASP {
-           // For testnet, so it should never expire. So set to u64::MAX
-           expiration_date: 18446744073709551615,
-           human_name,
-           base_url,
-           compliance_public_key,
-        }
+        move_to(
+            vasp,
+            ParentVASP {
+                // For testnet, so it should never expire. So set to u64::MAX
+                expiration_date: 18446744073709551615,
+                human_name,
+                base_url,
+                compliance_public_key,
+            }
+        )
+    }
+
+    /// Create a child VASP resource for the `parent`
+    /// Aborts if `parent` is not a ParentVASP
+    public fun publish_child_vasp_credential(parent: &signer, child: &signer) {
+        let parent_vasp_addr = Signer::address_of(parent);
+        Transaction::assert(exists<ParentVASP>(parent_vasp_addr), 7000);
+        move_to(child, ChildVASP { parent_vasp_addr });
     }
 
     ///////////////////////////////////////////////////////////////////////////
     // Publicly callable APIs
     ///////////////////////////////////////////////////////////////////////////
 
-    // Create a child VASP resource for the sender
-    // Aborts if the sender is not a ParentVASP
-    public fun create_child_vasp(sender: &signer): ChildVASP {
-        ChildVASP { parent_vasp_addr: Signer::address_of(sender) }
+    /// Return `addr` if `addr` is a `ParentVASP` or its parent's address if it is a `ChildVASP`
+    /// Aborts otherwise
+    public fun parent_address(addr: address): address acquires ChildVASP {
+        if (exists<ParentVASP>(addr)) {
+            addr
+        } else if (exists<ChildVASP>(addr)) {
+            borrow_global<ChildVASP>(addr).parent_vasp_addr
+        } else { // wrong account type, abort
+            abort(88)
+        }
     }
 
-    public fun child_parent_address(child: &ChildVASP): address {
-        child.parent_vasp_addr
+    public fun is_parent(addr: address): bool {
+        exists<ParentVASP>(addr)
     }
 
-    // Return true if `sender` is the parent VASP of `ChildVASP`
-    public fun is_parent_vasp(child_vasp: &ChildVASP, sender: &signer): bool {
-        Signer::address_of(sender) == child_vasp.parent_vasp_addr
+    public fun is_child(addr: address): bool {
+        exists<ChildVASP>(addr)
```

Comment:
We unfortunately can't because it would cause a circular dependency... But we can ask the prover folks to specify/prove a global invariant  that `exists<ChildVASP>(addr) <=> global<LibraAccount>(addr).role_id == CHILD_VASP_ROLE_ID`

Reply:
CC @DavidLDill